google/api_core/retry/retry_streaming.py

```diff
@@ -0,0 +1,330 @@
+        on_error: If given, the on_error callback will be called with each
+            retryable exception raised by the target. Any error raised by this
+            function will *not* be caught.
+        exception_factory: A function that is called when the retryable reaches
+            a terminal failure state, used to construct an exception to be raised.
+            It it given a list of all exceptions encountered, a retry.RetryFailureReason
+            enum indicating the failure cause, and the original timeout value
+            as arguments. It should return a tuple of the exception to be raised,
+            along with the cause exception if any.
+            If not provided, a default implementation will raise a RetryError
+            on timeout, or the last exception encountered otherwise.
+        init_args: Positional arguments to pass to the target function.
+        init_kwargs: Keyword arguments to pass to the target function.
+
+    Returns:
+        Generator: A retryable generator that wraps the target generator function.
+
+    Raises:
+        ValueError: If the sleep generator stops yielding values.
+        Exception: a custom exception specified by the exception_factory if provided.
+            If no exception_factory is provided:
+                google.api_core.RetryError: If the deadline is exceeded while retrying.
+                Exception: If the target raises an error that isn't retryable.
+    """
+
+    timeout = kwargs.get("deadline", timeout)
+    deadline: Optional[float] = (
+        time.monotonic() + timeout if timeout is not None else None
+    )
+    error_list: list[Exception] = []
+    # make a partial with timeout applied
+    exc_factory = lambda e, t: (exception_factory or _build_retry_error)(  # noqa: E731
+        e, t, timeout
+    )
+
+    for sleep in sleep_generator:
+        # Start a new retry loop
+        try:
+            # Note: in the future, we can add a ResumptionStrategy object
+            # to generate new args between calls. For now, use the same args
+            # for each attempt.
+            subgenerator = target(*init_args, **init_kwargs)
+            return (yield from subgenerator)
+        # handle exceptions raised by the subgenerator
+        # pylint: disable=broad-except
+        # This function explicitly must deal with broad exceptions.
+        except Exception as exc:
+            # defer to shared logic for handling errors
+            _retry_error_helper(
+                exc, deadline, sleep, error_list, predicate, on_error, exc_factory
+            )
+            # if exception not raised, sleep before next attempt
+            time.sleep(sleep)
+
+    raise ValueError("Sleep generator stopped yielding sleep values.")
+
+
+class StreamingRetry(_BaseRetry):
+    """Exponential retry decorator for streaming synchronous RPCs.
+
+    This class returns a Generator when called, which wraps the target
```

Comment:
If we decide to make this a breaking change, maybe we can remove/vastly simplify the whole discussion about deadline vs timeout? My impression was it was originally added to transition the `deadline` variable to `timeout`. If we have a new major version, we can probably just drop most references to deadline, since it has been deprecated for a while (Maybe just keep parsing it out of kwargs for compatibility, but not mention it in the signatures/docstrings?). Then we can just keep a simple explanation of what timeout specifically means

If not, keeping it in the `Retry` class sounds good to me

Reply:
That's a good point. We should decide about the breaking change issue.